### ISSUE TITLE
Issue on latest version Next12 & Cypress 10 - testing. Having same issue with Cypress 9 project too. 

### ISSUE THREAD

**Issue by ghost** [bonnie/udemy-NEXTJS-TESTING#3]
Hi Bonnie. When running project with Cypress 10 and NextJs 12.2 I have one or two issues.
This one <https://github.com/bonnie/udemy-NEXTJS-TESTING/tree/main/tested-concert-venue-nextjs12.2-cypress10.3>

Pre-setup.

I did as README. 
- Clone project
- use the commands to create env files. 

env file content is exposed here. This is the repo I have issues with. Do you know how to solve them? 
https://github.com/stephyswe/nextjs-testing/tree/main/tested-concert-venue-nextjs12.2-cypress10.3

One issue is running "npm test:ci" - the test runs fine one by one. But running the test suite gives an error on reservations.test.js - called "POST /api/reservations/[reservationId] creates a reservation". The test fail by showing expected length of 3. when sometimes the length of 2 comes along. Picture added. 

![image](https://user-images.githubusercontent.com/20955623/196375938-93e90063-e36b-42e0-aacf-67fa649d8859.png)

Another issue. Is with Cypress.  
- I run "npm run dev" to have client running. 
- Then I run "npm run cypress:open" 
- Then running tests "E2E Testing" - I get failure on auth-wrapper.
Picture added here too.

![image](https://user-images.githubusercontent.com/20955623/196376754-0a4b2319-157b-45e4-a2ac-d9b521e0e408.png)

**Comment by bonnie** [comment]
Hi, Stephanie, 

Thanks for reaching out! 

> One issue is running "npm test:ci" - the test runs fine one by one. But running the test suite gives an error on reservations.test.js - called "POST /api/reservations/[reservationId] creates a reservation". The test fail by showing expected length of 3. when sometimes the length of 2 comes along. Picture added.

I was able to reproduce this. I believe it's a race condition with test files running in parallel that are sharing the same test DB. One of the files resets the DB for a test at just the wrong time, so that the work of the test (POSTing a new reservation) is wiped out before the assertion runs. The possible solutions are: 

1. update the script to run only one file at a time: `"test:ci": "jest --ci --max-workers=1",`
  - advantages: relatively easy
  - disadvantages: makes tests run more slowly (in this case only adds half a second to one second, but in a large test suite it could be quite significant)
2. use a separate test database for each test file. This is more work, but if test speed is very important, it's the better way to go. 


> Another issue. Is with Cypress.
> 
>     I run "npm run dev" to have client running.
>     Then I run "npm run cypress:open"
>     Then running tests "E2E Testing" - I get failure on auth-wrapper.

I was able to reproduce this as well. 

This has to do with the database expectations of the Cypress tests. They need to be using the test database, and in order for that to happen, you need to be running the test build. If I run `npm run cypress:build` the error does not occur. 

Cheers,
Bonnie

**Comment by bonnie** [comment]
By the way, I discuss all of this (the parallel tests issue and how the test database works) in the course associated with this repository: [Testing Next.js Apps with Jest, Testing Library and Cypress](https://www.udemy.com/course/nextjs-testing/?couponCode=TEST-NEXTJS-2022-10%22). If you would like to support my work (these courses are how I make my living) and learn more about this repository, please consider purchasing the course. 

Thanks,
Bonnie

**Comment by ghost** [comment]
Thank you bonnie, you rock!
will buy your course next week, promise!
